### PR TITLE
Update to PHPUnit namespaces

### PR DESCRIPTION
**Pull request by peterrehm**
| Q             | A
| ------------- | ---
| Branch?       | 2.7+
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21534 
| License       | MIT
| Doc PR        | -

Replaces #21540

@nicolas-grekas Is the update of the cache-id like this sufficient? Do we maybe have to specifiy 4.8.35 in simple-phpunit?

### CODE REVIEW

**Comment by peterrehm** [comment]
The FC layer is basically restricted to a few classes as you can see here: https://github.com/sebastianbergmann/phpunit/commit/a55235c033c2dc18db631cc6b804844f6d37615a

- PHPUnit_Framework_Assert
- PHPUnit_Framework_BaseTestListener
- PHPUnit_Framework_TestCase
- PHPUnit_Framework_TestListener

In order to fully switch to namespaces we must upgrade to 6.0.

**Comment by peterrehm** [comment]
The PHP 5.3 test fails due to the wrong phpunit version.

**Comment by stof** [comment]
Well, you say you are replacing #21540, but our remarks on it are still valid. We have more public base test cases (and we have some listeners in the PHPUnit bridge)

**Comment by peterrehm** [comment]
@stof Yes. Sorry, it looks like I did not elaborate enough information about your Bridge remark.

The issue is since only the above mentioned 4 classes are in the FC layer, the Bridge cannot be updated to 6.0 as it uses the following classes:

````
/symfony/src/Symfony/Bridge/PhpUnit$ pt PHPUnit_
bootstrap.php
16:if (!defined('PHPUNIT_COMPOSER_INSTALL') && !class_exists('PHPUnit_TextUI_Command', false)) {

DeprecationErrorHandler.php
54:                return \PHPUnit_Util_ErrorHandler::handleError($type, $msg, $file, $line, $context);
62:            while (1 < $i && (!isset($trace[--$i]['class']) || ('ReflectionMethod' === $trace[$i]['class'] || 0 === strpos($trace[$i]['class'], 'PHPUnit_')))) {
76:                    || in_array('legacy', \PHPUnit_Util_Test::getGroups($class, $method), true)
99:            if (array('PHPUnit_Util_ErrorHandler', 'handleError') === $oldErrorHandler) {

Tests/DeprecationErrorHandler/default.phpt
21:class PHPUnit_Util_Test

TextUI/Command.php
17:class Command extends \PHPUnit_TextUI_Command
```` 

Are those the listeners/handlers you meant? I do not see a way other then adding classes to the bridge which should then work like the FC layer here https://github.com/sebastianbergmann/phpunit/commit/a55235c033c2dc18db631cc6b804844f6d37615a.

**Comment by peterrehm** [comment]
I would think all the other remarks are included in the PR.

**Comment by stof** [comment]
@peterrehm it cannot be switched to namespaces entirely to add compat, I know. But several places can be changed.

And we also have base test cases elsewhere than the KernelTestCase (I know that the Form and Validator components have some)

**Comment by peterrehm** [comment]
How about migrating at least ALL those 4 classes to the namespaced version that the FC layer provides?

File: src/Symfony/Bridge/ProxyManager/Tests/LazyProxy/ContainerBuilderTest.php [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
 require_once __DIR__.'/Fixtures/includes/foo.php';
 
+use PhpUnit\Framework\TestCase;
```
Comment (stof) [inline src/Symfony/Bridge/ProxyManager/Tests/LazyProxy/ContainerBuilderTest.php]: wrong namespace. It is ``PHPUnit`` (same in many files, and it breaks on the CI server)

Comment (peterrehm) [inline src/Symfony/Bridge/ProxyManager/Tests/LazyProxy/ContainerBuilderTest.php]: Fixed with the Migration of the Components

**Comment by peterrehm** [comment]
And `PHPUnit_Framework_Assert` is also migrated.

**Comment by peterrehm** [comment]
I dont get why the PHP 7 tests are failing, on my local machine I do not have any issues.
Any other remarks?

**Comment by peterrehm** [comment]
/ping @symfony-deciders Any remarks? Can we get this merged?

File: composer.json [hunk 2]
```diff
@@ -24,6 +24,9 @@
         "twig/twig": "~1.28|~2.0",
         "psr/log": "~1.0"
     },
+    "conflict": {
+        "phpunit/phpunit": "< 4.8.35 || >=5.0 <=5.4.2"
```
Comment (jderusse) [inline composer.json]: Symfony 2.7 supports php 5.3
This rule prevent people to install a version of phpunit compatible with php < 5.6

**Comment by Jean85** [comment]
If this gets merged, due to the new conflicts with the older PHPUnit versions, I can completely rewrite my PR for the PHPUnit bridge too (#21221)

File: composer.json (hunk 2 above)
Comment (peterrehm) [inline composer.json]: Should not as it allows 4.8.35?

**Comment by jderusse** [comment]
If this PR break unit tests for a supported version, you should find a way to fix it

**Comment by peterrehm** [comment]
@jderusse As mentioned I do not get why they are failing, it might be related to the rebuild of the bridge, the subtree split, etc. Feel free to take over if you know better.

File: phpunit [hunk 3]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env php
 <?php
 
-// Cache-Id: https://github.com/symfony/phpunit-bridge/commit/4f2aa873d2872a3b9782b25879fd628a0c418bc9
+// Cache-Id: https://github.com/symfony/phpunit-bridge/commit/d3157d942a4590121dfd23f9cadf519ca6ad4ac7
```
Comment (xabbuh) [inline phpunit]: This change is not necessary. The issue with the failing tests is that the PhpUnitBridge itself checks whether or not it needs an update. However, this doesn't work here as the `simple-phpunit` file is part of the bridge installed as a vendor library.

Comment (nicolas-grekas) [inline phpunit]: It still is necessary I think to trigger a phpunit cache invalidation on appveyor.

File: composer.json (hunk 2 above)
Comment (jderusse) [inline composer.json]: Indeed, I missed the strict comparison

**Comment by nicolas-grekas** [comment]
I cleared the phpunit cache and restarted tests. There are still a few failures to take care of.

**Comment by peterrehm** [comment]
Hm, when quickly scanning the logs I do not understand the reason for the failures. I see some missing autoloadphp's and some command failures.

Do you have a clue what the reason might be?

**Comment by nicolas-grekas** [comment]
Dunno yet about the autoload failure.
You should be able to reproduce (then debug) this one easily locally, isn't it?
https://travis-ci.org/symfony/symfony/jobs/199575966#L3178

**Comment by peterrehm** [comment]
Actually not at all:

````
With PHPUnit 5.7.
./phpunit src/Symfony/Bundle/TwigBundle/Tests/Functional/CacheWarmingTest.php
#!/usr/bin/env php
PHPUnit 5.7.12 by Sebastian Bergmann and contributors.

Testing Symfony\Bundle\TwigBundle\Tests\CacheWarmingTest
..                                                                  2 / 2 (100%)

Time: 430 ms, Memory: 14.00MB

With PHPUnit 4.8

./phpunit src/Symfony/Bundle/TwigBundle/Tests/Functional/CacheWarmingTest.php
#!/usr/bin/env php
PHPUnit 4.8.35 by Sebastian Bergmann and contributors.

Testing Symfony\Bundle\TwigBundle\Tests\CacheWarmingTest
..

Time: 456 ms, Memory: 14.00MB
````

I tried some adjustment, it just makes no sense. Localy the tests pass on PHP7 as well.

**Comment by peterrehm** [comment]
With the latest commits I fixed 5.3 which IMHO makes no sense but who cares.

No there are the issues with 7.x ledt which seem more related with the test setup rather than the actual changes? Is this maybe due to the subtree split?

File: composer.json (hunk 2 above)
Comment (nicolas-grekas) [inline composer.json]: should written as `<4.8.35|<=5.4.2,>=5.0` for consistency with what we do elsewhere IMHO

**Review by nicolas-grekas — APPROVED** [review]
LGTM - would be great to have a rule on fabbot that ensures that we won't introduce non-namespaced phpunit things anymore thought.

**Comment by nicolas-grekas** [comment]
travis failures are unrelated (known composer bug when too many local packages are created)

**Comment by peterrehm** [comment]
@nicolas-grekas Updated the constraint. For now some PHPUnit_ classes are still needed due to the limited scope of the FC layer.

If we manage to provide a full FC layer we could enforce that. Next would be to update the bridge to full PHP6 compatibility.

**Comment by nicolas-grekas** [comment]
Just did that: `grep PHPUnit_ * -r`, which leads to two things that remain:
all doc block annotations should be updated to namespaces, even if the BC layer doesn't provide them.
all instanceof/catch checks should be doubled with both namespaced and underscore versions.

**Comment by Jean85** [comment]
@nicolas-grekas what about the usages of static methods in the bridge, like those?

 * https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php#L73
 * https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/PhpUnit/DeprecationErrorHandler.php#L94

We don't have an FC layer for that..

**Comment by peterrehm** [comment]
See my attempt here: https://github.com/sebastianbergmann/phpunit/issues/2502 Lets see how it goes.

**Comment by Jean85** [comment]
@peterrehm already done that, was shut down: https://github.com/sebastianbergmann/phpunit/pull/2414

**Comment by peterrehm** [comment]
Well, I wanna go the other round, extending the existing Forward Copatibility Layer by a few classes.

**Comment by nicolas-grekas** [comment]
For the bridge, we should add a conflict header for phpunit6 in its composer.json file. phpunit6 compat should be only for master.
There are still many `@return \PHPUnit_Framework_MockObject_MockObject` or `@var \PHPUnit_Framework_MockObject_MockObject` and similar that should replaced by namespaced alternatives.

**Comment by peterrehm** [comment]
Added the conflict header. If that is okay with you I would like to change all the other occurences only once we have an actual solution for them. I hope we get the FC layer of PHPUnit extended.

**Comment by nicolas-grekas** [comment]
We already have one solution for the annotation: just use phpunit 6. There is no technical side effect to these annotations so we should change them right now - no need for any bc/fc layer for them.

**Comment by peterrehm** [comment]
If you want to, I will have a look at all the ones in comments. I just stumbled upon the MockObjects, looks like they are not yet namespaced:

https://github.com/sebastianbergmann/phpunit-mock-objects/blob/master/src/MockObject.php

The other dockblocks are updated.

**Comment by peterrehm** [comment]
@Jean85 Okay, got declined as well. That's a shame.

**Comment by nicolas-grekas** [comment]
PR or at least issue on phpunit-mock-objects to move to namespaces?

**Comment by peterrehm** [comment]
I am sure Sebastian is fully aware of this and we can adopt once there are changes.

However I opened up an issue: https://github.com/sebastianbergmann/phpunit-mock-objects/issues/353

File: src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php [hunk 4]
```diff
@@ -14,6 +14,7 @@
 use Doctrine\Common\Annotations\AnnotationReader;
 use Doctrine\ORM\Mapping\Driver\AnnotationDriver;
 use Doctrine\ORM\EntityManager;
+use PHPUnit\Framework\TestCase;
```
Comment (phansys) [inline src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php]: Actually, with the constraint  `<4.8.35|<5.4.3,>=5.0`, the real QCN is [`phpunit\framework\TestCase`](https://github.com/sebastianbergmann/phpunit/blob/5.4.2/src/ForwardCompatibility/TestCase.php#L11).

**Review by phansys — CHANGES_REQUESTED** [review]
Status: Needs work.

File: src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php (hunk 4 above)
Comment (juliendufresne) [inline src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php]: `<4.8.35|<5.4.3,>=5.0`  is in the conflict section of composer.json

What it really means is that we want PHPUnit 4.8.35 (the last version of PHPUnit 4) or a PHPUnit>=5.4.3
As you noticed, 5.4.3 is the first version of PHPUnit5 with the namespace ` PHPUnit\Framework\TestCase`

Comment (stof) [inline src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php]: We actually don't want 5.4.3, as the namespace was not the same (lowercase vs not lowercase). Use the first version using the ``PHPUnit`` case instead.

Comment (peterrehm) [inline src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php]: @stof Actually we want 5.4.3 as this is the first one with the correct namespace (https://github.com/sebastianbergmann/phpunit/blob/5.4.3/src/ForwardCompatibility/TestCase.php#L11). 5.4.2 had the lowercase one. https://github.com/sebastianbergmann/phpunit/blob/5.4.3/src/ForwardCompatibility/TestCase.php#L11

Comment (stof) [inline src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php]: ah indeed. What requires newer versions is the compat layer for more PHPUnit classes (which is only in 4.8.35 and 5.7). But I think this concerns only the bridge.

Comment (peterrehm) [inline src/Symfony/Bridge/Doctrine/Test/DoctrineTestHelper.php]: Yes, I wanted to look into this once this is merged. But so far it is pending.

**Comment by peterrehm** [comment]
@nicolas-grekas Is there anything preventing this getting merged?

**Comment by nicolas-grekas** [comment]
a rebase is needed - would you be able to try a PR on php-cs-fixer to add a rule so that we can detect and fix non namespaced uses in the future?

**Comment by peterrehm** [comment]
@nicolas-grekas Rebased. Once this is merged I would like to check first the 3.x branches then we should care about the bridge. Once we have this done we could look into how to prevent those issues. Does fabbot rely on php-cs-fixer?

**Comment by Jean85** [comment]
@peterrehm I this get merged, I would like to help with the bridhe.
I'll close and attempt a rewrite to my prev PR (#21221) but I think it'll be a hard think to do... The main issue that I was unable to overcome was the usage of the TestCase in argument typehints, and the use of the FC layer should be helpful!

**Comment by peterrehm** [comment]
@Jean85 Lets discuss once this is done. I think we have to look into providing a FC layer as PHPUnit does.

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by xabbuh** [comment]
:+1: The changes look good. I just don't understand why the build is failing.

**Comment by peterrehm** [comment]
@xabbuh https://github.com/symfony/symfony/pull/21564#issuecomment-279336727

**Comment by Tobion** [comment]
👍

**Review by Tobion — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @peterrehm.

**Comment by nicolas-grekas** [comment]
@peterrehm can you please check branch 2.8 now? I just merged 2.7 into 2.8. Thanks for your help.

**Pull request closed without merging**